### PR TITLE
[jit]maskrcnn & bert AD coverage part 1

### PR DESCRIPTION
**Pull request by ailzhang**
- Moved a few functions from `autograd` namespace to `aten` namespace to be visible from JIT nativeResolver.
- Added a hack to loop up keyword only argument. Will add proper support for kw only later
- Simulate function overload in aten using `_<number>` as function name suffix.
- Even `forward` returns multiple outputs like in `kthvalue`, there's at most one requires grad that we currently support. 
- Removed the `TensorList` related ops here since partial `TensorList` support is prone to bugs. Our symbolic diff for `cat` was never tested with autodiff, and it seems broken. Need to find another proper way to support these ops(either by properly supporting `TensorList` or sth like `prim::ConstantChunk`  and leave them for next PR.

Ops supported in this PR:
```
erf
expand_as
index
kthvalue
mean
permute
pow
rsub
select
sqrt
squeeze
t
to
topk
transpose
view
var
embedding
logsumexp
// grad is None
_dim_arange
contiguous
nonzero
ones_like
```

### CODE REVIEW

**Comment by ailzhang** [comment]
Failed tests are unrelated. Can I get a review on this @zdevito @apaszke ? Thanks!

File: torch/csrc/jit/symbolic_script.cpp [hunk 1]
```diff
@@ -114,6 +357,14 @@ c10::optional<GradientPair> gradientInfoForSchema(
     return cache_it->second;
   } else {
     auto schema_str = canonicalSchemaString(schema);
+    // JIT doesn't support keyword only arguments.
```
Comment (zdevito) [inline torch/csrc/jit/symbolic_script.cpp]: can you make a follow up issue to tracking adding support for `, *,` to the parser.

File: torch/csrc/jit/symbolic_script.cpp [hunk 2]
```diff
@@ -90,7 +323,17 @@ void loadModule(const std::shared_ptr<script::Module>& module) {
...
         loaded_schema.arguments(),
         {originalReturnType(new_tuple->type()->expect<TupleType>())});
+
+    // modify canonical string for function overloading
+    // prefer not to modify the schema name
+    const auto& schema_name = actual_schema.name();
+    auto pos = schema_name.find_last_of('_');
+    auto schema_name_suffix = schema_name.substr(pos + 1);
     std::string key = canonicalSchemaString(actual_schema);
+    if (!schema_name_suffix.empty() && schema_name_suffix.find_first_not_of("0123456789") == string::npos) {
```
Comment (zdevito) [inline torch/csrc/jit/symbolic_script.cpp]: Can you refactor the string processing into function?

**Review by zdevito — APPROVED** [review]
This looks good. I have a few nits, most important of which is to avoid saving `self` if all that is needed is `self.sizes()`. Saving something just for its sizes potentially uses a lot of memory. I tried to find all the places where just the sizes could be saved.

File: torch/csrc/jit/symbolic_script.cpp [hunk 3]
```diff
@@ -6,20 +6,253 @@ namespace {
...
+            return result0, result1, backward
+
+        def mean_0(self):
+            def backward(grad_output):
+                grad_self = grad_output.expand(self.size()) / self.numel()
+                return grad_self
+
+            return torch.mean(self), backward
+
+        def mean_1(self,
```
Comment (apaszke) [inline torch/csrc/jit/symbolic_script.cpp]: What is `mean_0` and `mean_1`? How do we match up those things with definitions later?

**Review by apaszke — COMMENTED** [review]
Why are we adding a bunch of backward ops instead of putting them as Python code in AD? From what I understand the plan is to deduplicate them later, and start using Python for everything, so it doesn't seem necessary to do that in every case

File: torch/csrc/jit/symbolic_script.cpp (hunk 3 above)
Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: These suffixes will be removed later when looking up function schema cache. This is a hack to simulate function overloading we have in aten. https://github.com/pytorch/pytorch/pull/16689/files/1a3d72ec52ddc4755008dfe920d4c0a25b8a1e13#diff-758cb565d2c84fc5268de0b63958f78dR330

**Comment by ailzhang** [comment]
@apaszke I added those backward function since they already existed in `autograd` namespace with correct behavior. so I tried avoid duplicating implementations. 

In general I think if we want Python for everything in future, maybe we can move them after we enrich our Torchscript with a larger subset of Python language? (eg. list comprehension etc).

**Comment by ailzhang** [comment]
And here are some perf numbers I got for `mean`. 
1. Testing for two small sizes, `(32, 32)` and `(32,)`
2. By commenting out `mean_0` and `mean_1` in `symbolic_script.cpp`, it's using `autograd` where `_safe_sizes` is in `autograd` namespace. 
Otherwise it's using AD defined using TorchScript where `_safe_sizes` is in `at` namespace.

| Autograd/ ms | forward | backward | 
|--------------|---------|----------|
| (32, 32)     | 0.26971 | 0.49648  |
| (32,)        | 0.26757 | 0.48329  |  


| TorchScript/ ms | forward | backward | 
|-----------------|---------|----------|
| (32, 32)        | 0.23459 | 0.45463  | 
| (32,)           | 0.23290 | 0.40245  | 

Attaching the script I used for benchmarking
```
import torch
import torch.nn as nn
import time

@torch.jit.script
def func(input):
    return torch.mean(input)
func.debug_disable_autodiff_subgraph_inlining()

def bm(testsize, repeat=1000, cuda=False):
    warmup = 10
    total_time = 0.0
    input = torch.rand(testsize, device='cuda' if cuda else 'cpu').float()
    input.requires_grad = True
    for _ in range(warmup):
        l = func(input)

    for _ in range(repeat):
        start = time.time()
        # l = func(input)
        l.backward()
        end = time.time()
        total_time += end - start
    return total_time

for cuda in [False, ]:
    for testsize in [(32, ),]:
        print('{:.5f}'.format(bm(testsize, cuda=cuda)))
```

**Comment by ailzhang** [comment]
Also attaching perf number for lstm benchmark here. 
### Without this PR

```
Namespace(device='cuda', hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=False, rnns=['cudnn', 'aten', 'jit'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)
Benchmarking LSTMs...
            name          avg_fwd          std_fwd          avg_bwd          std_bwd
           cudnn            5.184           0.2184            9.774           0.3062
            aten            5.734          0.01491            17.86           0.9224
             jit            6.277           0.1695            16.31           0.4344

Benchmarking ResNets...
            name          avg_fwd          std_fwd          avg_bwd          std_bwd
        resnet18            15.11           0.2244            31.62           0.4586
    resnet18_jit            15.28           0.3882            32.07           0.7986
        resnet50            48.09           0.6213            99.37           0.9248
    resnet50_jit            48.43           0.6271            99.95           0.8638

```

### With this PR

```

Namespace(device='cuda', hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=False, rnns=['cudnn', 'aten', 'jit'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)
Benchmarking LSTMs...
            name          avg_fwd          std_fwd          avg_bwd          std_bwd
           cudnn            5.142            0.164            9.514           0.2422
            aten            5.706          0.02338            16.63          0.04731
             jit            6.821           0.2923            16.07           0.9905

Benchmarking ResNets...
            name          avg_fwd          std_fwd          avg_bwd          std_bwd
        resnet18            15.02           0.2081            31.49           0.3507
    resnet18_jit            15.31           0.4251            32.11           0.9156
        resnet50            48.07           0.6603            99.39            1.076
    resnet50_jit            48.37           0.6513             99.9           0.9768
```

**Comment by zdevito** [comment]
6.277 vs  6.821 is like an 8% slowdown. Is that number stable across runs, in which case it seems bad, or is it just noise? The test of the mean operator should be compared against the base revision as well since we want to know if we regressed the backward speed by adding the aten op.

File: torch/csrc/jit/symbolic_script.cpp (hunk 3 above)
Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: - name: mean(Tensor self)




- name: mean(Tensor self, ScalarType dtype)
- name: mean(Tensor self, IntList dim, bool keepdim)
- name: mean(Tensor self, IntList dim, ScalarType dtype)
- name: mean(Tensor self, IntList dim, bool keepdim, ScalarType dtype)

There are mean overloads that has dtype in the signature, should we also include those overloads?

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: @wanchaol Yea but this `dtype` option is not exposed in our public api. Thus these ops are not specifically tested in common_methods_invocations.py. Do you know some ops that call into the dtype version of mean? 
I didn't want to add some ops without proper tests, thus those were not in this PR.

Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: I found one callsites in [here](https://github.com/pytorch/pytorch/blob/41dddfd55fc59a0ebae7cda7e759b80011741a78/test/test_cuda.py#L2017), but that's only a test and no ops found that calls `mean` with dtype, so I guess it should be fine.

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: aha this option is actually exposed but not documented :( `a = torch.mean(a, dtype=torch.double)` works. I'm can add AD & test for them.

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: Actually I realized all dtypes here are keyword only arguments, and since we don't document them I assume it's rarely used in real models :P 
I would prefer to add them together with proper keyword-only argument parser so that we are less hacky on this. But if there's a usecase that needs this AD formula, let me know so that I can prioritize. :D

Comment (wanchaol) [reply torch/csrc/jit/symbolic_script.cpp]: I don't think we have too much use cases on this either. Let's just add this with the keyword arg parser together.

**Comment by ailzhang** [comment]
My old benchmarks were run with `DEBUG=1` and the timings vary a lot among runs. Deleting... From my current observation with `cset`, the resnet benchmark is pretty stable and consistent with master branch. 
JIT run time for lstm benchmark is volatile itself as on master branch, it ranges from 6.0 - 6.3 in 5 runs with a large variance 0.26. I noticed with this PR in, it's noticeably a little slower than master branch as the range move to 6.1 - 6.4 in 5 runs etc. But with this benchmark, it's pretty hard to debug where this slowness came from.

File: torch/csrc/jit/symbolic_script.cpp [hunk 4]
```diff
@@ -51,6 +354,24 @@ Argument originalReturnType(const TupleTypePtr& tup) {
...
+// overloaded functions of `func`.
+// Remove the suffix before adding the schema string to map
+// schema_to_graphs.
+std::string overloadedSchemaString(const FunctionSchema& schema) {
+  const auto& schema_name = schema.name();
+  auto pos = schema_name.find_last_of('_');
+  auto schema_name_suffix = schema_name.substr(pos + 1);
+  std::string schema_string = canonicalSchemaString(schema);
+  if (!schema_name_suffix.empty()
+      && schema_name_suffix.find_first_not_of("0123456789") == string::npos) {
```
Comment (sinkingsugar) [inline torch/csrc/jit/symbolic_script.cpp]: missing `std::` before `string::npos`, failing here on all my builds :/, weird that no CI catched this. @ailzhang

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: @sinkingsugar  Would you mind using the collect_env.py (https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py) to let me know your environment? 
Basically this is indeed a typo on my side, but it got hidden by this header file included indirectly.  https://github.com/pytorch/pytorch/blob/master/c10/util/string_utils.h#L6 We likely want to fix this on our side.  But I'm curious why your build failed. Thanks!

Comment (ezyang) [reply torch/csrc/jit/symbolic_script.cpp]: At a guess, probably some ifdefs toggled differently causing the header to not be included.

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: @ezyang Yea agree, would love to know @sinkingsugar 's build environment, maybe worth adding to our CI?

Comment (ezyang) [reply torch/csrc/jit/symbolic_script.cpp]: We can probably figure it out just by tracing the include path ourselves :)

Comment (sinkingsugar) [reply torch/csrc/jit/symbolic_script.cpp]: @ailzhang 
well, didn't have time to run the script but I will just give you the full Dockerfile 😄 
https://gist.github.com/sinkingsugar/81a42c6104bf01efd00e1d4ff9358c9b

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: @sinkingsugar Interestingly the Dockerfile built successfully on my end :( 
screen shot 2019-02-20 at 11 05 43 am

Comment (sinkingsugar) [reply torch/csrc/jit/symbolic_script.cpp]: Hmm weird :) pretty sure I gave you the right one! but on master it's now fixed anyway so all good I suppose!
